repl/src/main/scala/ammonite/repl/interp/Interpreter.scala

```diff
@@ -311,12 +328,70 @@ class Interpreter(prompt0: Ref[String],
       (classFiles, newImports) <- compiled
       _ = Timer("cachedCompileBlock 4")
       cls <- eval.loadClass(fullyQualifiedName, classFiles)
-    } yield (cls, newImports)
+    } yield (cls, newImports, tag)
+  }
+
+
+  def evalCachedClassFiles(cachedData: Seq[ClassFiles],
+                           pkg: String,
+                           wrapper: String): Res[Seq[Unit]] = {
+
+    def getBlockNumber(blockNumber: Int) = blockNumber match {
+      case 1 => ""
+      case _ => "_" + blockNumber.toString
+    }
+
+      val cls = Res.map(cachedData.zipWithIndex)(
+        (data: (ClassFiles, Int)) => {
+          val (clsFiles, index) = data
+          Compiler.addToClasspath(clsFiles, dynamicClasspath)
+          eval.loadClass(pkg + "." + wrapper + getBlockNumber(index + 1), clsFiles)
+        }
+      )
+      cls.map(_.map(eval.evalMain(_)))
+
   }
 
-  def processModule(source: ImportHook.Source, code: String, wrapperName: Name, pkgName: Seq[Name]) = {
 
-    processModule0(source, code, wrapperName, pkgName, predefImports)
+  def processModule(source: ImportHook.Source,
+                    code: String,
+                    wrapperName: Name,
+                    pkgName: Seq[Name]
+                   ): Res[Imports] = {
+    val cacheTag = Util.md5Hash(Iterator(code.getBytes)).map("%02x".format(_)).mkString
+    storage.classFilesListLoad(
+      pkgName.map(_.backticked).mkString("."),
+      wrapperName.backticked,
+      cacheTag
+    ) match {
+      case None =>
+        init()
+        val res = processModule0(source, code, wrapperName, pkgName, predefImports)
+        res match{
+         case Res.Success(data) =>
+           reInit()
+           val (imports, files) = data
+           storage.classFilesListSave(
+             pkgName.map(_.backticked).mkString("."),
+             wrapperName.backticked,
+             files,
+             imports,
+             cacheTag
+           )
+           res.map(_._1)
+         case r: Res.Failing => r
+       }
+      case Some((cachedData, imports)) =>
```

Comment:
You are failing to cache import hooks, and thus any necessary magic imports like `import $file.foo` to load secondary files or Ivy dependencies will not run on cached scripts, which will cause failures.